### PR TITLE
Use getaddrinfo to parse IPv6 addresses to allow link local addresses

### PR DESCRIPTION
**Pull request by fcrisciani**
Commit Message: use only `getaddrinfo` to parse IPv6
Risk Level: low
Testing: unit tests

### CODE REVIEW

File: test/common/network/utility_test.cc [hunk 1]
```diff
@@ -150,8 +153,10 @@ TEST(NetworkUtility, ParseInternetAddress) {
   EXPECT_EQ(nullptr, Utility::parseInternetAddressNoThrow("0:0:0:0"));
   EXPECT_EQ(nullptr, Utility::parseInternetAddressNoThrow("fffff::"));
   EXPECT_EQ(nullptr, Utility::parseInternetAddressNoThrow("/foo"));
+#ifndef WIN32
   EXPECT_EQ(nullptr, Utility::parseInternetAddressNoThrow("[::]"));
   EXPECT_EQ(nullptr, Utility::parseInternetAddressNoThrow("[::1]:1"));
+#endif
```
Comment (yanavlasov) [inline test/common/network/utility_test.cc:159]: I think in the NoThrow case it makes sense to have the `#else` clause to show what happens on Windows.

**Review by yanavlasov — DISMISSED** [review]
/wait

File: test/common/network/utility_test.cc (hunk 1 above)
Comment (fcrisciani) [reply test/common/network/utility_test.cc:162]: done.

Comment (wrowe) [reply test/common/network/utility_test.cc:162]: Same logic above at 142 (note an extra empty line there before #if), below at 237 & 251?

We might want to record a bug to track a different PR to the getaddrinfo call wherein our wrapper enforces a little more strictness, feel free to assign me to it.

**Review by wrowe — CHANGES_REQUESTED** [review]
Otherwise makes sense to me

File: test/common/network/utility_test.cc (hunk 1 above)
Comment (fcrisciani) [reply test/common/network/utility_test.cc:164]: created the TODO with the bug number and added an extra space before the compiler conditions.

**Review by wrowe — CHANGES_REQUESTED** [review]
Thanks for creating the issue! My original request wasn't as much for consistent whitespace (but that's also goodness, TY), but to illustrate all the exceptions so we can document the broken behavior and proceed to fix them in OsSysCalls, as another PR.

File: test/common/network/utility_test.cc [hunk 2]
```diff
@@ -139,8 +139,11 @@ TEST(NetworkUtility, ParseInternetAddress) {
   EXPECT_THROW(Utility::parseInternetAddress("0:0:0:0"), EnvoyException);
   EXPECT_THROW(Utility::parseInternetAddress("fffff::"), EnvoyException);
   EXPECT_THROW(Utility::parseInternetAddress("/foo"), EnvoyException);
+
+#ifndef WIN32
   EXPECT_THROW(Utility::parseInternetAddress("[::]"), EnvoyException);
   EXPECT_THROW(Utility::parseInternetAddress("[::1]:1"), EnvoyException);
+#endif
```
[the remark this answers is not in the document]
Comment (fcrisciani) [reply test/common/network/utility_test.cc:146]: I'm happy to add it but it's the same as line 160-164, do you want to add here and everywhere else the same parsing issue?

**Pull request merged**